---
title: "dropRunner Report"
output: 
  html_document:
    theme: flatly
date: "`r format(Sys.time(), '%d %B, %Y')`"
params:
  sampleID: ""
---

```{r setup, include=FALSE, fig.align='center'}
knitr::opts_chunk$set(echo = TRUE, fig.width = 6, fig.height = 4, fig.align = 'center')
```

```{r, echo=F}
suppressMessages(library(ggplot2))
suppressMessages(library(scales))
suppressMessages(library(kableExtra))
theme_set(theme_bw())

config <- read.delim('config.yaml',sep=':', header=F, stringsAsFactors = F)

human.solo.out<- read.delim(paste0(params$sampleID,'_human/Gene/Summary.csv'),sep=',', header=F, stringsAsFactors = F)
chimp.solo.out<- read.delim(paste0(params$sampleID,'_chimp/Gene/Summary.csv'),sep=',', header=F, stringsAsFactors = F)

human.UMIperCell <- read.delim(paste0(params$sampleID,'_human/Gene/UMIperCellSorted.txt'), sep=',', header=F, stringsAsFactors = F)
chimp.UMIperCell <- read.delim(paste0(params$sampleID,'_chimp/Gene/UMIperCellSorted.txt'), sep=',', header=F, stringsAsFactors = F)

#picard <- read.delim(paste0('output/qc_data/',params$sampleID,'_RNAmetrics.picard.txt'), skip = 6)[1,]

```

# Summary

```{r, echo=F}
n <- as.integer(human.solo.out[10,'V2'])
meanReads <- as.integer(human.solo.out[13,'V2'])
project <- config[8,'V2']
protocol <- config[9, 'V2']
```
#### Protocol: $\color{blue}{\text{`r protocol`}}$
#### Project Name: $\color{blue}{\text{`r project`}}$
#### Sample: $\color{blue}{\text{`r params$sampleID`}}$
#### Estimated number of cells: $\color{green}{\text{`r n`}}$
#### Mean Reads per Cell: $\color{green}{\text{`r meanReads`}}$

***

## Sequencing

### Human

```{r, echo=F, fig.align='center'}
seq.summary <- human.solo.out[c(1:7),]
seq.summary[2:7,'V2'] <- paste0(round(seq.summary[2:7,'V2']*100, digits = 1), '%')
colnames(seq.summary) <- c('Features','Statistics')
knitr::kable(seq.summary,row.names = F,caption = 'Sequencing summary') %>% kable_styling(bootstrap_options = c("striped", "hover"), full_width = F)
```

### Chimp

```{r, echo=F, fig.align='center'}
seq.summary <- chimp.solo.out[c(1:7),]
seq.summary[2:7,'V2'] <- paste0(round(seq.summary[2:7,'V2']*100, digits = 1), '%')
colnames(seq.summary) <- c('Features','Statistics')
knitr::kable(seq.summary,row.names = F,caption = 'Sequencing summary') %>% kable_styling(bootstrap_options = c("striped", "hover"), full_width = F)
```

***

## Cells

### Human

```{r, echo=F, fig.align='center'}
cell.summary <- human.solo.out[c(10,12,13,17,18,20),]
cell.summary[2,'V2'] <- paste0(round(cell.summary[2,'V2']*100, digits = 1), '%')
colnames(cell.summary) <- c('Features','Statistics')
knitr::kable(cell.summary,row.names = F,caption = 'Summary of cells detected') %>% kable_styling(bootstrap_options = c("striped", "hover"), full_width = F)
```

```{r, echo=F}
n <- as.integer(cell.summary[1,'Statistics'])
human.UMIperCell$r <- 1:nrow(human.UMIperCell)
human.UMIperCell$type <- ifelse(human.UMIperCell$r <= n, 'grey','black')
ggplot(human.UMIperCell, aes(x=r, y=V1, color=type)) + geom_point() + 
  theme(text = element_text(size=14), legend.position = 'none') +
  scale_y_continuous(trans = log10_trans(),
    breaks = trans_breaks("log10", function(x) 10^x),
    labels = trans_format("log10", math_format(10^.x))) +
  scale_x_continuous(trans = log10_trans(),
    breaks = trans_breaks("log10", function(x) 10^x),
    labels = trans_format("log10", math_format(10^.x))) +
  scale_color_manual(values=c('grey','black')) +
  xlab('Barcodes') + ylab('UMIs Counts')
```

### Chimp 

```{r, echo=F, fig.align='center'}
cell.summary <- chimp.solo.out[c(10,12,13,17,18,20),]
cell.summary[2,'V2'] <- paste0(round(cell.summary[2,'V2']*100, digits = 1), '%')
colnames(cell.summary) <- c('Features','Statistics')
knitr::kable(cell.summary,row.names = F,caption = 'Summary of cells detected') %>% kable_styling(bootstrap_options = c("striped", "hover"), full_width = F)
```

```{r, echo=F}
n <- as.integer(cell.summary[1,'Statistics'])
chimp.UMIperCell$r <- 1:nrow(chimp.UMIperCell)
chimp.UMIperCell$type <- ifelse(chimp.UMIperCell$r <= n, 'grey','black')
ggplot(chimp.UMIperCell, aes(x=r, y=V1, color=type)) + geom_point() + 
  theme(text = element_text(size=14), legend.position = 'none') +
  scale_y_continuous(trans = log10_trans(),
    breaks = trans_breaks("log10", function(x) 10^x),
    labels = trans_format("log10", math_format(10^.x))) +
  scale_x_continuous(trans = log10_trans(),
    breaks = trans_breaks("log10", function(x) 10^x),
    labels = trans_format("log10", math_format(10^.x))) +
  scale_color_manual(values=c('grey','black')) +
  xlab('Barcodes') + ylab('UMIs Counts')
```

***

## Species assigment

Distribution of human genome specificity

```{r}
assign_species <- read.delim(paste0(params$sampleID,'/species.csv'),sep='\t', header=T, stringsAsFactors = F)
hist(assign_species$hg_specificity_score, xlab = 'hg Specificity Score', ylab='Counts', main = 'Distribution of Human Specificity')
```

***

```{r, echo=F, eval=F}
genic <- picard$PCT_CODING_BASES + picard$PCT_UTR_BASES + picard$PCT_INTRONIC_BASES
exonic <-  picard$PCT_CODING_BASES + picard$PCT_UTR_BASES
intronic <- picard$PCT_INTRONIC_BASES
intergenic <- picard$PCT_INTERGENIC_BASES
pct_cov <- as.data.frame(rbind(genic, exonic, intronic, intergenic)) 
colnames(pct_cov) <- "pct_coverage"
pct_cov$feature <- c("genic", "exonic", "intronic", "intergenic")

ggplot(pct_cov, aes(feature, pct_coverage, label = pct_coverage)) +
  geom_bar(stat = "Identity", position = "dodge", fill = "lightblue") +
  scale_x_discrete(limits=pct_cov$feature) +
  ylab("Fraction of Reads") +
  xlab("") +
  geom_text(size = 3, position = position_stack(vjust = 0.5)) + 
  theme(text = element_text(size=14),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank())
```


